Software/Program/Code/Core/Src/main.c

```diff
@@ -90,18 +147,121 @@ int main(void)
+      if ( ADC_ConversionInProgress == false )
+      {
+         (void)HAL_ADC_Start( &hadc3 );   // TODO: Use status return value
+         ADC_ConversionInProgress = true;
+         HAL_ADC_PollForConversion(&hadc3, 80);
+      //}
+      //else if ( ADC_ConversionComplete == true )
+      //{
+         ADC_ConversionComplete = false;
+         ADC_ConversionInProgress = false;
+         adc_reading = HAL_ADC_GetValue( &hadc3 );
+         if ( adc_reading != Previous_ADCValue )
+         {
+            UpdateOutput = true;
+            Previous_ADCValue = adc_reading;
+         }
+      }
+
+      /*** HANDLE FREQUENCY OUTPUT ***/
+      if ( UpdateOutput == true )
+      {
+         uint32_t high_pulse_time_arr_value;
+         float period_us;
+         uint32_t period_in_timer_counts;
+         
+         UpdateOutput = false;
+
+         // Convert this ADC reading into a frequency...
+         Frequency = Interpolate_ADCCount_To_Frq( adc_reading );
+         period_us = 1.0e9 / (float)Frequency;
+         period_in_timer_counts = (uint32_t) ( period_us * TIMER_COUNTS_PER_NANOSECOND );
+
+         // Figure out new high-pulse time
+         switch( Direction )
+         {
+            case DIRECTION_FORWARD_MIN:
+               high_pulse_time_arr_value = (uint32_t) ( ( DIRECTION_FORWARD_HIGH_PULSE_TIME_NOMINAL - DIRECTION_FORWARD_HIGH_PULSE_TIME_TOLERANCE ) * TIMER_COUNTS_PER_NANOSECOND );
+               break;
+
+            case DIRECTION_FORWARD_NOMINAL:
+               high_pulse_time_arr_value = (uint32_t) ( DIRECTION_FORWARD_HIGH_PULSE_TIME_NOMINAL * TIMER_COUNTS_PER_NANOSECOND );
+               break;
+
+            case DIRECTION_FORWARD_MAX:
+               high_pulse_time_arr_value = (uint32_t) ( ( DIRECTION_FORWARD_HIGH_PULSE_TIME_NOMINAL + DIRECTION_FORWARD_HIGH_PULSE_TIME_TOLERANCE ) * TIMER_COUNTS_PER_NANOSECOND );
+               break;
+
+            case DIRECTION_REVERSE_MIN:
+               high_pulse_time_arr_value = (uint32_t) ( ( DIRECTION_REVERSE_HIGH_PULSE_TIME_NOMINAL - DIRECTION_REVERSE_HIGH_PULSE_TIME_TOLERANCE ) * TIMER_COUNTS_PER_NANOSECOND );
+               break;
+
+            case DIRECTION_REVERSE_NOMINAL:
+               high_pulse_time_arr_value = (uint32_t) ( DIRECTION_REVERSE_HIGH_PULSE_TIME_NOMINAL * TIMER_COUNTS_PER_NANOSECOND );
+               break;
+
+            case DIRECTION_REVERSE_MAX:
+               high_pulse_time_arr_value = (uint32_t) ( ( DIRECTION_REVERSE_HIGH_PULSE_TIME_NOMINAL + DIRECTION_REVERSE_HIGH_PULSE_TIME_TOLERANCE ) * TIMER_COUNTS_PER_NANOSECOND );
+               break;
+
+            case DIRECTION_ERROR:
+               high_pulse_time_arr_value = (uint32_t) ( DIRECTION_ERROR_HIGH_PULSE_TIME * TIMER_COUNTS_PER_NANOSECOND );
+               break;
+
+            default:
+               high_pulse_time_arr_value = (uint32_t) ( DIRECTION_ERROR_HIGH_PULSE_TIME * TIMER_COUNTS_PER_NANOSECOND );
+               break;
+         }
```

Comment:
Similar comment as above regarding the ADC → frequency setpoint. This section of code should be with the button press.